### PR TITLE
(dev) Pre-generate crypto materials

### PR DESCRIPTION
**Pull request by AurelienGasser**
Companion PR: https://github.com/SubstraFoundation/substra-tests/pull/151

# Overview 

Pre-generate HLF crypto materials to speedup local development:

- no enrollment pod
- no genesis pod
- orderer / peer pods start instantly

# Benchmarks (local env)

|  | **Total time to readiness** | skaffold run | Deployment stabilized | Chaincode instantiated | skaffold delete | 
|--|--|--|--|--|--|
| Without pre-gen | **143s**  | 100s | 39s | 43s | 25s |
| With pre-gen | **80s** | 60s | 5s | 20s | 13s | 

Note: on the CI, there's a big speedup in deployment stabilization time (from ~6 min to ~2 min)

# How to use

### Without pre-gen

```bash
$ skaffold run
```

### With pre-gen

```bash
$ ./examples/secrets-create.sh
$ skaffold run
```

## Delete

```
$ skaffold delete
```

Secrets will be deleted regardless of pre-gen status. To re-deploy with pre-gen, you need to run `examples/secrets-create.sh` again.

To delete secrets only, run `./examples/secrets-create.sh delete`

# Limitations and Future work

Skaffold deployment order is not configurable: by default, helm charts are deployed _before_ kubectl manifest.

So if the user doesn't  run `examples/secrets-create.sh` before `skaffold run`, skaffold will still create the secrets, but **after** deploying the helm charts. This greatly limits the speedup benefits. 

There is a [skaffold issue](https://github.com/GoogleContainerTools/skaffold/issues/3905) opened on the topic.  Once it's addressed, we can get rid of `examples/secrets-create.sh` altogether.

### CODE REVIEW

File: charts/hlf-k8s/templates/deployment-enrollement-operator.yaml [hunk 1]
```diff
@@ -167,3 +168,4 @@ spec:
       nodeSelector:
         {{- toYaml . | nindent 8 }}
       {{- end }}
+{{- end}}
```
Comment (inalgnu) [inline charts/hlf-k8s/templates/deployment-enrollement-operator.yaml:171]: missing EOF, if you are using VSCode this option might help
```
"files.insertFinalNewline": true
```

File: skaffold.yaml [hunk 2]
```diff
@@ -51,19 +51,19 @@ deploy:
...
-  # We create the serviceAccounts with the kubectl deployer to ensure the serviceAccounts will stay present for the Helm deletion hooks
-  # By design Skaffold create and delete ressources in a fixed order (Helm > kubectl > kustomize)
-  # Ref: https://github.com/GoogleContainerTools/skaffold/blob/dedd545/pkg/skaffold/runner/new.go#L191-L214 (func getDeployer)
+
+
   kubectl:
     manifests:
-      - examples/serviceAccounts/serviceAccount-orderer.yaml
-      - examples/serviceAccounts/serviceAccount-org-1.yaml
-      - examples/serviceAccounts/serviceAccount-org-2.yaml
```
Comment (Kelvin-M) [inline skaffold.yaml:69]: Why do you remove the serviceAccounts part, we don't need it when we use secrets instead of CA enrollment + genesis creation ?

**Comment by Kelvin-M** [comment]
I found `./examples/secrets-create.sh delete` pretty strange, maybe it would be better to have something like : 

```
./examples/hlf-secrets.sh create
./examples/hlf-secrets.sh delete
```

**Comment by Kelvin-M** [comment]
Moreover, we should warn the user that's only for dev setup and should not be used in production :D !

File: skaffold.yaml (hunk 2 above)
Comment (AurelienGasser) [reply skaffold.yaml:69]: Correct, we don't need them with pre-gen secrets. They're only used by the secret deletion hook

Here's the diff between the non-pregen/pregen orderer values

```diff
>   enabled: false   // that's the CA: now disabled
<   serviceAccount:   // no need for a service account
<     name: substra-delete-hook
>   deleteSecrets:     // that's the hook: now disabled
>     enabled: false
>
> genesis:               // the genesis creation: now disabled
>   generate: false
```

**Comment by AurelienGasser** [comment]
@Kelvin-M What about

```
./examples/sample-secrets.sh create
./examples/sample-secrets.sh delete
```

Would that address both your concerns?

File: charts/hlf-k8s/templates/deployment-enrollement-operator.yaml (hunk 1 above)
Comment (AurelienGasser) [reply charts/hlf-k8s/templates/deployment-enrollement-operator.yaml]: Fixed https://github.com/SubstraFoundation/hlf-k8s/pull/95/commits/45f3c3237f2ba2b6e5ad1d8769b5355547f5ca30

Thanks for the tip!

**Comment by AurelienGasser** [comment]
or `dev-secrets.sh` ?

**Review by Kelvin-M — APPROVED** [review]

**Pull request merged**